File: gaussian_elimination.py
import numpy as np


def back_substitution(A, b):
    for line in range(len(A)-1, -1, -1):
        if abs(A[line][line]) < 1e-12:
            print("Error")
            return
        S = 0
        for column in range(line+1, len(A), 1):
            S = S + A[line][column]*b[column]
        b[line] = (b[line] - S)/A[line][line]

    return b


def gaussian_elimination(A, b):
    n = len(A)-1
    for step in range(0, n, 1):
        if abs(A[step][step]) < 1e-12:
            print("Error")
            return
        for line in range(step+1, n+1, 1):
            m = A[line][step]/A[step][step]
            A[line][step] = 0
            b[line] = b[line] - m*b[step]
            for column in range(step+1, n+1, 1):
                A[line][column] = A[line][column] - m*A[step][column]

    return back_substitution(A, b)


# Example
if __name__ == '__main__':
    A = np.array([[2, 1, 1],
                  [4, 3, 3],
                  [8, 7, 9]], np.float64)

    b = np.array([[1],
                  [2],
                  [6]], np.float64)
    print(f'x = {gaussian_elimination(A, b)}')

File: src/Gaussian Elimination/gaussian_elimination.jl
function back_substitution(A::Array{Float64,2}, b::Array{Float64,1})
  n = length(b)

  for line = n:-1:1
    if abs(A[line, line]) <= 1e-12
      error("Matriz singular")
    end
    S = 0.0
    for column = line+1:1:n
      S = S + A[line, column] * b[column]
    end
    b[line] = (b[line] - S) / A[line, line]
  end
  return b
end


function gaussian_elimination(A::Array{Float64,2}, b::Array{Float64,1})
  n = length(b)
  for step = 1:1:n-1
    if abs(A[step, step]) <= 1e-12
      error("Erro")
    end
    for line = step+1:1:n
      m = A[line, step]/A[step, step]
      A[line, step] = 0
      b[line] = b[line] - m*b[step]
      for column = step+1:1:n
        A[line, column] = A[line, column] - m*A[step, column]
      end
    end
  end
  return back_substitution(A, b)
end

# Example
A = [2 1 1; 4 3 3; 8 7 9.0]
b = [1; 2; 6.0]
x = gaussian_elimination(A, b)

display(x)